components/Buttons/src/MDCFloatingButton+Animation.m

```diff
@@ -89,110 +86,190 @@ + (float)fab_dragCoefficient {
+    NSAssert(NO, @"The FAB should not be expanded unless its X and Y scales are less than 1.");
+    return;
+  }
+
   void (^expandActions)(void) = ^{
-    self.layer.transform =
-        CATransform3DConcat(self.layer.transform, [MDCFloatingButton expandTransform]);
-    self.layer.opacity = 1;
-    self.imageView.layer.transform =
-        CATransform3DConcat(self.imageView.layer.transform, [MDCFloatingButton expandTransform]);
-    [self.layer removeAnimationForKey:kMDCFloatingButtonTransformKey];
-    [self.layer removeAnimationForKey:kMDCFloatingButtonOpacityKey];
-    [self.imageView.layer removeAnimationForKey:kMDCFloatingButtonTransformKey];
+    [self.layer removeAnimationForKey:kMDCFloatingButtonTransformExpandKey];
+    [self.layer removeAnimationForKey:kMDCFloatingButtonOpacityVisibleKey];
+    [self.imageView.layer removeAnimationForKey:kMDCFloatingButtonTransformExpandKey];
     if (completion) {
       completion();
     }
   };
 
+  CATransform3D layerTransformToValue =
+      CATransform3DConcat(self.layer.transform, [MDCFloatingButton expandTransform]);
+
+  CALayer *iconPresentationLayer = self.imageView.layer.presentationLayer;
+  CATransform3D iconTransformToValue =
+      iconPresentationLayer ? iconPresentationLayer.transform : self.imageView.layer.transform;
+
   if (animated) {
     [CATransaction begin];
     [CATransaction setDisableActions:YES];
     [CATransaction setCompletionBlock:expandActions];
 
+    CALayer *presentationLayer = self.layer.presentationLayer;
+    NSValue *fromValue =
+        presentationLayer ? [NSValue valueWithCATransform3D:presentationLayer.transform] : nil;
+
+    CFTimeInterval duration =
+        [MDCFloatingButton fab_animationDurationFromWallTimeDuration:kMDCFloatingButtonEnterDuration
+                                                            forLayer:self.layer];
+    CFTimeInterval beginTime =
+        [MDCFloatingButton fab_animationBeginTimeFromWallTimeOffset:0 forLayer:self.layer];
+
     CABasicAnimation *overallScaleAnimation = [MDCFloatingButton
         animationWithKeypath:@"transform"
-                     toValue:[NSValue
-                                 valueWithCATransform3D:CATransform3DConcat(
-                                                            self.layer.transform,
-                                                            [MDCFloatingButton expandTransform])]
-                   fromValue:nil
+                     toValue:[NSValue valueWithCATransform3D:layerTransformToValue]
+                   fromValue:fromValue
               timingFunction:[[CAMediaTimingFunction alloc]
                                  initWithControlPoints:0.0f:0.0f:0.2f:1.0f]
-                    fillMode:kCAFillModeForwards
-                    duration:kMDCFloatingButtonEnterDuration
-                 beginOffset:0];
-    [self.layer addAnimation:overallScaleAnimation forKey:kMDCFloatingButtonTransformKey];
+                    fillMode:kCAFillModeBoth
```

Comment:
In a follow-up change we may want to document the name of this timing function (looks like a material easing curve?)

Reply:
It was a custom curve I got from the designers. If we have a repository of such curves, I could add it.